### PR TITLE
Added login functionality, refactored SignUpServlet

### PR DESCRIPTION
**Pull request by Greengrer**
1. Added SignInServlet and SignInServletTest.
2. Added signInCheck(String email, String password) to
   DefaultUserService.
3. Refactored add(User user) -> add(String email, String password)
   method in DefaultUserService.
4. Added DefaultUserServiceITest testing methods signInCheck and add.
5. Refactored SignUpServlet and SignUpServletTest.

### CODE REVIEW

File: src/main/java/com/vinylteam/vinyl/service/impl/DefaultUserService.java [hunk 1]
```diff
@@ -1,27 +1,56 @@
...
 import com.vinylteam.vinyl.entity.User;
+import com.vinylteam.vinyl.security.impl.DefaultSecurityService;
 import com.vinylteam.vinyl.service.UserService;
 
 import java.util.Optional;
 
 public class DefaultUserService implements UserService {
 
     private final JdbcUserDao jdbcUserDao;
+    private final DefaultSecurityService defaultSecurityService;
```
Comment (AlexKievDeveloper) [inline src/main/java/com/vinylteam/vinyl/service/impl/DefaultUserService.java]: Can we use interfaces instead implementations?

File: src/main/java/com/vinylteam/vinyl/service/impl/DefaultUserService.java [hunk 2]
```diff
@@ -1,27 +1,56 @@
...
 
-    public DefaultUserService() {
-        this.jdbcUserDao = new JdbcUserDao();
+    public DefaultUserService(JdbcUserDao jdbcUserDao, DefaultSecurityService defaultSecurityService) {
```
Comment (AlexKievDeveloper) [inline src/main/java/com/vinylteam/vinyl/service/impl/DefaultUserService.java]: The same situation: .... (UserDao userDao, SecurityService securityService) {

File: src/main/java/com/vinylteam/vinyl/web/servlets/SignInServlet.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+import org.slf4j.LoggerFactory;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.util.Optional;
+
+public class SignInServlet {
+
+    private final DefaultUserService defaultUserService;
```
Comment (AlexKievDeveloper) [inline src/main/java/com/vinylteam/vinyl/web/servlets/SignInServlet.java]: UserService userService;

File: src/main/java/com/vinylteam/vinyl/web/servlets/SignUpServlet.java [hunk 4]
```diff
@@ -11,29 +10,21 @@
...
-                    .createUserWithHashedPassword(email, password.toCharArray());
-
-            if (defaultUserService.add(userToAdd)) {
-                response.setStatus(HttpServletResponse.SC_OK);
-                response.sendRedirect("/signIn");
-                return;
-            }
+        if (defaultUserService.add(email, password)) {
+            response.setStatus(HttpServletResponse.SC_SEE_OTHER);
+            response.sendRedirect("/verify");
```
Comment (AlexKievDeveloper) [inline src/main/java/com/vinylteam/vinyl/web/servlets/SignUpServlet.java]: What will be after redirect to /verify? What servlet will process the /verify request?

File: src/main/java/com/vinylteam/vinyl/web/servlets/SignInServlet.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        String password = request.getParameter("password");
+
+        SignInCheckResult checkResult = defaultUserService.signInCheck(email, password);
+
+        if (checkResult == SignInCheckResult.OK_VERIFIED) {
+            response.setStatus(HttpServletResponse.SC_OK);
+            response.sendRedirect("/home");
+        } else if (checkResult == SignInCheckResult.OK_NOT_VERIFIED) {
+            response.setStatus(HttpServletResponse.SC_SEE_OTHER);
+            response.sendRedirect("/verify");
```
Comment (AlexKievDeveloper) [inline src/main/java/com/vinylteam/vinyl/web/servlets/SignInServlet.java]: What will be after redirect to /verify? What servlet will process the /verify request?

File: src/test/java/com/vinylteam/vinyl/web/servlets/SignUpServletTest.java [hunk 6]
```diff
@@ -17,107 +15,56 @@
...
-    private final User mockedUser = mock(User.class);
 
     @BeforeEach
     void beforeEach() {
-        mockedDefaultSecurityService = null;
         mockedDefaultUserService = null;
         mockedHttpServletRequest = null;
         mockedHttpServletResponse = null;
 
         mockedDefaultUserService = mock(DefaultUserService.class);
```
Comment (AlexKievDeveloper) [inline src/test/java/com/vinylteam/vinyl/web/servlets/SignUpServletTest.java]: Why mockedDefaultUserService = null; and then mockedDefaultUserService = mock(DefaultUserService.class);  ?
Can you write just mockedDefaultUserService = mock(DefaultUserService.class);?

File: src/main/java/com/vinylteam/vinyl/service/impl/DefaultUserService.java (hunk 2 above)
Comment (Greengrer) [reply src/main/java/com/vinylteam/vinyl/service/impl/DefaultUserService.java]: Oooooh, missed that.

File: src/main/java/com/vinylteam/vinyl/web/servlets/SignInServlet.java (hunk 3 above)
Comment (Greengrer) [reply src/main/java/com/vinylteam/vinyl/web/servlets/SignInServlet.java]: Yep.

File: src/test/java/com/vinylteam/vinyl/web/servlets/SignUpServletTest.java (hunk 6 above)
Comment (Greengrer) [reply src/test/java/com/vinylteam/vinyl/web/servlets/SignUpServletTest.java]: I'll try and see if verification works right.

File: src/main/java/com/vinylteam/vinyl/service/impl/DefaultUserService.java [hunk 7]
```diff
@@ -1,27 +1,56 @@
...
+                if (securityService.checkPasswordAgainstUserPassword(
+                        optionalUser.get(), password.toCharArray())) {
+                    if (optionalUser.get().getStatus()) {
+                        return SignInCheckResult.OK_VERIFIED;
+                    } else {
+                        return SignInCheckResult.OK_NOT_VERIFIED;
+                    }
+                }
+            }
+
```
Comment (AlexKievDeveloper) [inline src/main/java/com/vinylteam/vinyl/service/impl/DefaultUserService.java]: Please remove empty string

File: src/test/java/com/vinylteam/vinyl/dao/jdbc/mapper/UserRowMapperTest.java [hunk 8]
```diff
@@ -33,9 +35,12 @@ void mapFilledRowTest() throws SQLException {
         assertEquals("salt", actualUser.getSalt());
         assertEquals(1, actualUser.getIterations());
         assertEquals(Role.USER, actualUser.getRole());
+        assertTrue(actualUser.getStatus());
     }
 
     @Test
+
```
Comment (AlexKievDeveloper) [inline src/test/java/com/vinylteam/vinyl/dao/jdbc/mapper/UserRowMapperTest.java]: Please remove empty row

**Review by AlexKievDeveloper — DISMISSED** [review]

File: src/main/java/com/vinylteam/vinyl/web/servlets/SignUpServlet.java [hunk 9]
```diff
@@ -11,29 +10,21 @@
 public class SignUpServlet {
 
     private final DefaultUserService defaultUserService;
-    private final DefaultSecurityService defaultSecurityService;
 
-    public SignUpServlet(DefaultUserService defaultUserService,
-                         DefaultSecurityService defaultSecurityService) {
+    public SignUpServlet(DefaultUserService defaultUserService) {
```
Comment (TarasOrlovskyi) [inline src/main/java/com/vinylteam/vinyl/web/servlets/SignUpServlet.java]: I think the interface should also be used here

File: src/test/java/com/vinylteam/vinyl/service/impl/DefaultUserServiceITest.java [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+import static org.junit.jupiter.api.Assertions.*;
+
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+class DefaultUserServiceITest {
+
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+    private final DefaultSecurityService defaultSecurityService = new DefaultSecurityService();
+    private final JdbcUserDao jdbcUserDao = new JdbcUserDao();
+    private final DefaultUserService defaultUserService = new
+            DefaultUserService(jdbcUserDao, defaultSecurityService);
```
Comment (TarasOrlovskyi) [inline src/test/java/com/vinylteam/vinyl/service/impl/DefaultUserServiceITest.java]: should be interface:
DefaultSecurityService
DefaultUserService

File: src/test/java/com/vinylteam/vinyl/web/servlets/SignInServletTest.java [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+
+    private DefaultUserService mockedDefaultUserService;
+    private HttpServletRequest mockedHttpServletRequest;
+    private HttpServletResponse mockedHttpServletResponse;
+
+    @BeforeEach
+    void beforeEach() {
+        mockedDefaultUserService = null;
+        mockedHttpServletRequest = null;
+        mockedHttpServletResponse = null;
```
Comment (TarasOrlovskyi) [inline src/test/java/com/vinylteam/vinyl/web/servlets/SignInServletTest.java]: why are we writing this? Maybe the next 3 lines will be enough?

mockedDefaultUserService = mock(DefaultUserService.class);
mockedHttpServletRequest = mock(HttpServletRequest.class);
mockedHttpServletResponse = mock(HttpServletResponse.class);

**Review by TarasOrlovskyi — DISMISSED** [review]
i hope that you consider my comments

**Review by AlexKievDeveloper — APPROVED** [review]

File: src/test/java/com/vinylteam/vinyl/web/servlets/SignUpServletTest.java [hunk 12]
```diff
@@ -17,107 +15,56 @@
...
     private HttpServletRequest mockedHttpServletRequest;
     private HttpServletResponse mockedHttpServletResponse;
-    private final User mockedUser = mock(User.class);
 
     @BeforeEach
     void beforeEach() {
-        mockedDefaultSecurityService = null;
         mockedDefaultUserService = null;
         mockedHttpServletRequest = null;
         mockedHttpServletResponse = null;
```
Comment (TarasOrlovskyi) [inline src/test/java/com/vinylteam/vinyl/web/servlets/SignUpServletTest.java]: i think that we should delete this

**Review by TarasOrlovskyi — APPROVED** [review]

**Review by TarasOrlovskyi — APPROVED** [review]

**Pull request merged**